File: LeetCode/Practice_for_interview/Search/DFS/200.cpp
class Solution {
public:
    vector<pair<int, int>> directions = {{0,1}, {0,-1}, {-1, 0}, {1, 0}};
    int dfs(vector<vector<char>>& g, int x, int y) {
        if( x < 0 || x >= g.size() || y < 0 || y >= g[0].size() || g[x][y] == '0') {
            return 0;
        }
        int area = 1;
        g[x][y] = '0';
        for (auto d: directions) {
            area += dfs(g, x + d.first, y + d.second);
        }
        return area;
    }

    int numIslands(vector<vector<char>>& grid) {
        vector<int> islands;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == '1') {
                    int t = dfs(grid, i, j);
                    islands.emplace_back(t);
                }
            }
        }
        return islands.size();
    }
};

File: LeetCode/Practice_for_interview/Search/DFS/200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(x, y):
            if x < 0 or x >= m or y < 0 or y >= n or grid[x][y] == '0':
                return 0
            area = 1
            grid[x][y] = '0'
            
            for xi, yi in directions:
                area += dfs(x+xi, y+yi)
            
            return area
        
        directions = [[0,1], [0,-1], [-1,0], [1,0]]
        islands = 0
        m, n = len(grid), len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    t = dfs(i, j)
                    islands += 1
        return islands



from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # visited m x n
        # T: O(mn), S: O(mn)
        def dfs(x, y):
            if grid[x][y] == '0' or visited[x][y]:
                return
            visited[x][y] = 1
            
            for xi, yi in ([-1,0],[1,0],[0, -1], [0, 1]):
                if 0 <= x + xi < m and 0 <= y + yi < n:
                    dfs(x + xi, y + yi)
    
        m, n = len(grid), len(grid[0])
        visited = [[0 for _ in range(n)] for _ in range(m)]
        # print(visited)
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1' and not visited[i][j]:
                    dfs(i, j)
                    res += 1
        return res

grid = [["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]
s1 = Solution()
print(s1.numIslands(grid))